e2e/main.sh

```diff
@@ -0,0 +1,219 @@
+    echo "   - Both hosts already exchanged their public keys:"
+    echo "     - Test runner has MicroShift's sshd key in ~/.ssh/known_keys"
+    echo "     - Remote \$USHIFT_USER has test runner's key in ~/.ssh/authorized_keys"
+    echo "   - Passwordless sudo for \$USHIFT_USER"
+    echo ""
+    echo " Script aims to be target platform agnostic. It means that for some environments (e.g. GCP)"
+    echo " it might be required to export firewall::open_port and firewall::close_port functions"
+    echo " so the tests can open custom ports"
+
+    [ -n "$1" ] && echo -e "\nERROR: $1"
+    exit 1
+}
+
+log() {
+    echo -e "$(date +'%H:%M:%S.%N')    $*"
+}
+
+ssh_cmd() {
+    local cmd="${1}"
+    ssh -o BatchMode=yes "${USHIFT_USER}@${USHIFT_IP}" "${cmd}"
+}
+
+var_should_not_be_empty() {
+    local var_name=${1}
+    if [ -z "${!var_name+x}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is unset"
+        return 1
+    elif [ -z "${!var_name}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is empty"
+        return 1
+    fi
+}
+
+function_should_be_exported() {
+    local fname=${1}
+    if ! declare -F "${fname}"; then
+        log "WARNING: Function '${fname}' is unexported. It is expected that function is provided for interacting with cloud provider"
+        return 1
+    fi
+}
+
+check_passwordless_ssh() {
+    ssh_cmd "true" || {
+        echo "Failed to access ${USHIFT_IP}:"
+        echo "  - Test runner should have MicroShift's sshd key in ~/.ssh/known_keys"
+        echo "  - Remote \$USHIFT_USER should have test runner's key in ~/.ssh/authorized_keys"
+        exit 1
+    }
+}
+
+check_passwordless_sudo() {
+    ssh_cmd "sudo --non-interactive true" || {
+        echo "Failed to run sudo command as ${USHIFT_USER} without password"
+        exit 1
+    }
+}
+
+prechecks() {
+    var_should_not_be_empty USHIFT_IP && var_should_not_be_empty USHIFT_USER || exit 1
+    check_passwordless_ssh
+    check_passwordless_sudo
+
+    # Just warning for now
+    # Following functions needed only for runs in CI
+    function_should_be_exported firewall::open_port || true
+    function_should_be_exported firewall::close_port || true
+}
+
+microshift_get_konfig() {
+    tmpfile=$(mktemp /tmp/microshift-e2e-konfig.XXXXXX)
+    ssh_cmd 'sudo cat /var/lib/microshift/resources/kubeadmin/'"${USHIFT_IP}"'/kubeconfig' >"${tmpfile}"
+    echo "${tmpfile}"
+}
+
+microshift_check_readiness() {
+    local output_dir="${1}"
+    log "Waiting for MicroShift to become ready"
+    ssh_cmd "sudo /etc/greenboot/check/required.d/40_microshift_running_check.sh" &>"${output_dir}/0002-readiness-check.log"
+}
+
+microshift_setup() {
+    local output_dir="${1}"
+    log "Setting up and starting MicroShift"
+    ssh_cmd 'cat << EOF | sudo tee /etc/microshift/config.yaml
+---
+apiServer:
+  subjectAltNames:
+  - '"${USHIFT_IP}"'
+EOF' &>"${output_dir}/0001-setup.log"
```

Comment:
Is this needed? Should it remain as part of the installation?

Reply:
But it's only done in CI, missing for local VMs (unless someone does it manually). I'd prefer using IPs over hostnames if we can

Reply:
Tests only require a host for the execution, be that a hostname or an IP, right? I would put the requirement on configuring the system outside of the tests so that they are generic for any environment.
In CI you have that as part of the install steps, and in a dev environment you may get away with default configuration. For example, I use "microshift-rhel9" as hostname for my VM and never use its IP, therefore the default config is ok.